Sources/NIO/CircularBuffer.swift

```diff
@@ -79,6 +85,87 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
         }
     }
 
+    /// Replaces the specified subrange of elements with the given collection.
+    ///
+    /// - Parameter subrange:
+    /// The subrange of the collection to replace. The bounds of the range must be valid indices of the collection.
+    ///
+    /// - Parameter newElements:
+    /// The new elements to add to the collection.
+    ///
+    /// *O(n)* where _n_ is the lenght of the new elements collection if the subrange equals to _n_
+    ///
+    /// *O(m)* where _m_ is the combined length of the collection and _newElements_
+    public mutating func replaceSubrange(_ subrange: RangeType<Int>, with newElements: ContiguousArray<E>) {
+        guard subrange.lowerBound >= self.startIndex && subrange.upperBound <= self.endIndex else {
+            preconditionFailure("Subrange out of bounds")
+        }
+
+        let lowerBound = bufferIndex(ofIndex: subrange.lowerBound)
+        let upperBound = bufferIndex(ofIndex: subrange.upperBound == self.endIndex ? subrange.upperBound - 1 : subrange.upperBound)
+
+        if subrange.count == newElements.count {
+            for (index, element) in zip(subrange.indices, newElements) {
+                self.buffer[self.bufferIndex(ofIndex: index)] = element
+            }
+        } else if subrange.count == self.count && newElements.isEmpty {
+            self.removeSubrange(subrange)
+        } else {
+            var newBuffer: ContiguousArray<E?> = []
+            newBuffer.reserveCapacity(self.buffer.count)
+            var index = self.headIdx
+            while index != self.tailIdx {
+                if index == lowerBound {
+                    for element in newElements {
+                        newBuffer.append(element)
+                    }
+                    while index != upperBound {
+                        index = self.index(after: index)
```

Comment:
I *think* you're mixing index types here. I think `headIdx` and `tailIdx` are the internal indices, not the public ones. Instead you probably want `startIndex` and `endIndex`.